### PR TITLE
[FIX] EditDomain: Editing TimeVariable broke formatting

### LINKED ISSUE #2776: Edit Domain destroys datetime format
##### Orange version

3.8.dev

##### Expected behavior
Edit Domain retains the set datetime iso format, e.g. 2017-11-14 12:12.

##### Actual behavior
Edit Domain converts datetime to floats, creating a mess downstream.

##### Steps to reproduce the behavior
Use on any datetime data set.

screen shot 2017-11-14 at 13 40 05

##### Additional info (worksheets, data, screenshots, ...)

Original data
screen shot 2017-11-14 at 13 39 49

After Edit Domain
screen shot 2017-11-14 at 13 39 57

### PR DESCRIPTION
**Pull request by robertcv**
##### Issue
Fixes #2776 

##### Description of changes
Add TimeVariableEditor and update variable_description and variable_from_description to correctly handle TimeVariables.

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

### CODE REVIEW

File: Orange/widgets/data/oweditdomain.py [hunk 1]
```diff
@@ -56,27 +56,39 @@ def variable_description(var, skip_attributes=False):
...
     except (ImportError, AttributeError):
         raise ValueError("Invalid descriptor type '{}.{}"
                          "".format(module, type_name))
 
-    var = constructor(name, compute_value=compute_value, **dict(list(kwargs)))
+    kwargs = dict(list(args))
+    var = constructor(name, compute_value=compute_value, **kwargs)
     var.attributes.update(attrs)
+    if var.is_time:
+        var.have_date = kwargs['have_date']
```
Comment (astaric) [inline Orange/widgets/data/oweditdomain.py]: Have you considered adding have_date and have_time to TimeVariable constructor as keyword arguments?

File: Orange/widgets/data/tests/test_oweditdomain.py [hunk 2]
```diff
@@ -169,6 +170,33 @@ def test_duplicate_names(self):
...
+    def test_time_variable_preservation(self):
+        """Test if time variables preserve format specific attributes"""
+        table = Table("cyber-security-breaches")
+        self.send_signal(self.widget.Inputs.data, table)
+        output = self.get_output(self.widget.Outputs.data)
+        self.assertEqual(str(table[0, 4]), str(output[0, 4]))
+
+        # if continuous variable edit is used have_time and have_date is not
+        # copied and the time string format is changed
+        idx = self.widget.domain_view.model().index(4)
```
Comment (astaric) [inline Orange/widgets/data/tests/test_oweditdomain.py]: Can you try to use https://github.com/biolab/orange3/blob/master/Orange/widgets/tests/utils.py#L273 here?

File: Orange/widgets/data/tests/test_oweditdomain.py [hunk 3]
```diff
@@ -169,6 +170,33 @@ def test_duplicate_names(self):
...
+        self.widget.domain_view.setCurrentIndex(idx)
+        editor = self.widget.editor_stack.findChild(ContinuousVariableEditor)
```
Comment (astaric) [inline Orange/widgets/data/tests/test_oweditdomain.py]: Not sure why this is tested her. Can ContinuousVariableEditor be used for editing TimeVariable by users?

File: Orange/widgets/data/tests/test_oweditdomain.py [hunk 4]
```diff
@@ -221,3 +249,20 @@ def test_discrete_editor(self):
...
+        w.set_data(v)
+
+        self.assertEqual(w.name_edit.text(), v.name)
+        self.assertEqual(w.labels_model.get_dict(), v.attributes)
+        self.assertTrue(w.is_same())
+
+        w.set_data(None)
+        self.assertEqual(w.name_edit.text(), "")
+        self.assertEqual(w.labels_model.get_dict(), {})
+        self.assertIs(w.get_data(), None)
```
Comment (astaric) [inline Orange/widgets/data/tests/test_oweditdomain.py]: You might want to add some tests that assure that the new variable has the same value for have_date and have_time as source.

File: Orange/widgets/data/oweditdomain.py [hunk 5]
```diff
@@ -508,11 +546,13 @@ def open_editor(self, index):
 
         var = self.domain_model[index]
 
-        editor_index = 2
+        editor_index = 3
         if var.is_discrete:
             editor_index = 0
-        elif var.is_continuous:
+        elif var.is_time:
```
Comment (astaric) [inline Orange/widgets/data/oweditdomain.py]: Can you add a comment why time variables need to be handled before numeric variables?

File: Orange/widgets/data/oweditdomain.py [hunk 6]
```diff
@@ -56,27 +56,39 @@ def variable_description(var, skip_attributes=False):
                 (("values", tuple(var.values)),),
                 attributes)
     else:
-        return (var_type.__module__,
-                var_type.__name__,
-                var.name,
-                (),
-                attributes)
+        if var.is_time:
```
[the remark this answers is not in the document]
Comment (robertcv) [reply Orange/widgets/data/oweditdomain.py]: No, it would be really better to use elif.

File: Orange/widgets/data/oweditdomain.py [hunk 7]
```diff
@@ -412,6 +449,7 @@ def __init__(self):
         self.editor_stack = QStackedWidget()
 
         self.editor_stack.addWidget(DiscreteVariableEditor())
+        self.editor_stack.addWidget(TimeVariableEditor())
```
Comment (astaric) [inline Orange/widgets/data/oweditdomain.py]: I would add this after ContinuousVariableEditor.

this way, the next diff should register as an addition of `elif var.is_time` as the `elif var.is_continuous` branch could remain unmodified

File: Orange/widgets/data/oweditdomain.py (hunk 5 above)
Comment (robertcv) [reply Orange/widgets/data/oweditdomain.py]: TimeVariable is also continuous so we need to test if_time first. Will add comment.

File: Orange/widgets/data/oweditdomain.py (hunk 1 above)
Comment (robertcv) [reply Orange/widgets/data/oweditdomain.py]: I was considering this but was afreide of breking something else, will add have_time and have_date as keyword arguments to Tiemariable constructor.

**Comment by astaric** [comment]
@kernc, are you ok with adding have_date and have_time kwargs to TimeVariable? They make construction of new variables where we already know what kind of datetime they contain much easier.

**Comment by kernc** [comment]
Was always in favor of setting parameters through the constructor. :+1:

**Pull request merged**